src/flask_wtf/csrf.py

```diff
@@ -77,42 +77,112 @@ def validate_csrf(data, secret_key=None, time_limit=None, token_key=None):
 
     :raises ValidationError: Contains the reason that validation failed.
 
+    .. versionchanged:: 1.21.post1
+        Fallbacks to legacy_validate_csrf method. This provides a
+        comptability layer for old clients.
     .. versionchanged:: 0.14
         Raises ``ValidationError`` with a specific error message rather than
         returning ``True`` or ``False``.
     """
+    try:
+        secret_key = _get_config(
+            secret_key,
+            "WTF_CSRF_SECRET_KEY",
+            current_app.secret_key,
+            message="A secret key is required to use CSRF.",
+        )
+        field_name = _get_config(
+            token_key,
+            "WTF_CSRF_FIELD_NAME",
+            "csrf_token",
+            message="A field name is required to use CSRF.",
+        )
+        time_limit = _get_config(time_limit, "WTF_CSRF_TIME_LIMIT", 3600, required=False)
 
-    secret_key = _get_config(
-        secret_key,
-        "WTF_CSRF_SECRET_KEY",
-        current_app.secret_key,
-        message="A secret key is required to use CSRF.",
-    )
-    field_name = _get_config(
-        token_key,
-        "WTF_CSRF_FIELD_NAME",
-        "csrf_token",
-        message="A field name is required to use CSRF.",
-    )
-    time_limit = _get_config(time_limit, "WTF_CSRF_TIME_LIMIT", 3600, required=False)
+        if not data:
+            raise ValidationError("The CSRF token is missing.")
+
+        if field_name not in session:
+            raise ValidationError("The CSRF session token is missing.")
 
-    if not data:
-        raise ValidationError("The CSRF token is missing.")
+        s = URLSafeTimedSerializer(secret_key, salt="wtf-csrf-token")
 
-    if field_name not in session:
-        raise ValidationError("The CSRF session token is missing.")
+        try:
+            token = s.loads(data, max_age=time_limit)
+        except SignatureExpired as e:
+            raise ValidationError("The CSRF token has expired.") from e
+        except BadData as e:
+            raise ValidationError("The CSRF token is invalid.") from e
+
+        if not hmac.compare_digest(session[field_name], token):
+            raise ValidationError("The CSRF tokens do not match.")
+    except Exception as e:
```

Comment:
> if I'm reading correctly, those will just never succeed (with old or new scheme), so we can just early-exit, right?

I think so

 are you looking for something like:
```py
try:
  ...
except ValidationError as e:
        if e.message == "The CSRF token is missing.":
            raise e
        if e.message == "The CSRF session token is missing.":
            raise e
        logger.info("Falling back to legacy CSRF validation.")
        token_key = 'csrf_token' if token_key is None else token_key
        is_valid = legacy_validate_csrf(
            data=data,
            secret_key=secret_key,
            time_limit=time_limit,
            token_key=token_key
        )
        if is_valid is False:
            raise e  
```

I think this works and saves us the call to `legacy_validate_csrf`, but I'd say it complicates the fork, because we're adding more code / more conditional to tests. Just more places for bugs to sneak in.



> this broad except is hard to reason about;

Even if we add early exit handling for `ValidationError`, I think I want to keep the broad except around for the `token = s.loads(data, max_age=time_limit)` line. I haven't tested, but I could imagine the old CSRF signed token could cause this function to raise an unhandled error -- but we'd still want that data to be tried against the `legacy_validate_csrf()` function